crates/turborepo-repository/src/change_mapper.rs

```diff
@@ -8,37 +8,85 @@ use wax::Program;
 
 use crate::package_graph::{ChangedPackagesError, PackageGraph, PackageName, WorkspacePackage};
 
+/// Detects if a file is in a package. If no package is detected, returns `None`
+pub trait PackageDetector {
+    fn detect_package(&self, file: &AnchoredSystemPath) -> Option<WorkspacePackage>;
+}
+
+/// Detects package by checking if the file is inside the package.
+///
+/// NOTE: This strategy has some limitations. Since it doesn't
+/// check the inputs field in `turbo.json`, any inputs that
+/// are not contained in the package directory won't be mapped
+/// to the correct package. Also, since we don't have the global dependencies,
+/// any file that is not in any package will automatically return the root
+/// package. This is fine for Vercel builds, but less fine for situations like
+/// watch mode.
+pub struct DefaultPackageDetector<'a> {
+    pkg_dep_graph: &'a PackageGraph,
+}
+
+impl<'a> DefaultPackageDetector<'a> {
+    pub fn new(pkg_dep_graph: &'a PackageGraph) -> Self {
+        Self { pkg_dep_graph }
+    }
+    fn is_file_in_package(file: &AnchoredSystemPath, package_path: &AnchoredSystemPath) -> bool {
+        file.components()
+            .zip(package_path.components())
+            .all(|(a, b)| a == b)
+    }
+}
+
+impl<'a> PackageDetector for DefaultPackageDetector<'a> {
+    fn detect_package(&self, file: &AnchoredSystemPath) -> Option<WorkspacePackage> {
+        for (name, entry) in self.pkg_dep_graph.packages() {
+            if name == &PackageName::Root {
+                continue;
+            }
+            if let Some(package_path) = entry.package_json_path.parent() {
+                if Self::is_file_in_package(file, package_path) {
+                    return Some(WorkspacePackage {
+                        name: name.clone(),
+                        path: package_path.to_owned(),
+                    });
+                }
+            }
+        }
+
+        Some(WorkspacePackage::root())
+    }
+}
+
 // We may not be able to load the lockfile contents, but we
 // still want to be able to express a generic change.
 pub enum LockfileChange {
     Empty,
     WithContent(Vec<u8>),
 }
 
+#[derive(Debug, PartialEq)]
```

Comment:
Is there a reason for not also implementing `Eq`?

Reply:
Not really, I just added what I needed